message/infrastructure/http/pubsub_test.go

```diff
@@ -0,0 +1,44 @@
+package http
+
+import (
+	"testing"
+	"time"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/stretchr/testify/require"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/ThreeDotsLabs/watermill/message/infrastructure"
+)
+
+func createPubSub(t *testing.T) message.PubSub {
+	logger := watermill.NewStdLogger(true, true)
+
+	sub, err := NewSubscriber(":8080", DefaultUnmarshalMessageFunc, logger)
+
+	go func() {
+		err := sub.StartHTTPServer()
+		require.NoError(t, err)
+	}()
+
+	time.Sleep(200 * time.Millisecond)
```

Comment:
From what I've read, `ListenAndServe` is blocking and there is no way to check when it's ready. The `Sleep` was ugly and hopefully, temporary. [Here](https://grokbase.com/t/gg/golang-nuts/141ee4dqyg/go-nuts-how-to-know-when-listenandserve-is-ready-to-handle-connections#201401153vyc35xvn3kqhbzrfvg2rqdivy) they say that you can call `Listen("tcp", addr)` and then `go http.Serve()` and be sure that the server is already listening. But `Subscriber` doesn't expose `Listen`, only `ListenAndServe`. However, I don't feel too good about exposing `Listen` as well, because it's a bit low-level. Maybe I could rewrite `Subscriber.ListenAndServe` to use this hack.

Reply:
but isn't the http client himself waiting for the server when it's not available?

Reply:
I've had the test fail because of a race condition. I've removed the sleep already, but i've had to modify `ListenAndServe` (`Listen(), go Serve, return`) to prevent a case where there would be nowhere to send the request to.